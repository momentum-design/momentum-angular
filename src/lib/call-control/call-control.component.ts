/** @component call-control */

import {
  Component,
  ElementRef,
  EventEmitter,
  Input,
  OnInit,
  Output,
} from '@angular/core';

export type CallControlSize =
| 'activities'
| 'camera'
| 'camera-muted'
| 'cancel'
| 'handset'
| 'microphone-muted'
| 'more'
| 'participant-list'
| 'share-screen'
| 'speaker'
| 'view-list';

const activeRed = ['camera-muted', 'cancel', 'microphone-muted', 'speaker'];
const activeBlue = ['camera', 'participant-list', 'share-screen'];

@Component({
  selector: 'md-call-control',
  template: `
    <button
      md-button
      [attr.aria-label]="ariaLabel || type"
      [attr.id]="id"
      [circle]="true"
      (click)="onClick($event)"
      [color]="type === 'cancel' ? 'red' : getColor()"
      [disabled]="disabled"
      [ngClass]="['md-call-control', hostClassName]"
      [size]="size"
    >
      <md-icon [name]="type + '_' + iconSize" [color]="iconColor"></md-icon>
    </button>
  `
})
export class CallControlComponent implements OnInit {
  /** @prop Sets active css styling | false */
  @Input() active: boolean = false;
  /** @prop Text to display for blindness accessibility features | '' */
  @Input() ariaLabel: string = '';
  /** @prop Sets the attribute disabled to the CallControl button | false */
  @Input() disabled: boolean = false;
  /** @prop Optional icon color prop | $md-white-100 */
  @Input() iconColor: string = 'white-100';
  /** @prop Optional numeric icon size prop | 24 */
  @Input() iconSize: number = 24;
  /** @prop Optional id string | null */
  @Input() id: string;
  /** @prop Optional numeric size prop for CallControl button | 56 */
  @Input() size: number | string = 56;
  /** @prop Optional predefined CallControl prop type | null */
  @Input() type: CallControlSize;

  @Output() readonly click: EventEmitter<MouseEvent> = new EventEmitter<MouseEvent>();

  hostClassName: string;

  constructor(private _element: ElementRef) {}

  ngOnInit() {
    this.hostClassName = this._element.nativeElement.className;
    const nativeElement: HTMLElement = this._element.nativeElement;
    const parentElement: HTMLElement = nativeElement.parentElement;
    while (nativeElement.firstChild) {
      parentElement.insertBefore(nativeElement.firstChild, nativeElement);
    }
    parentElement.removeChild(nativeElement);
  }

  private _getActiveColor() {
    switch (true) {
      case(activeRed.includes(this.type)):
        return 'red';
      case(activeBlue.includes(this.type)):
        return 'blue';
      default:
        return 'dark-gray';
    }
  }

  getColor() {
    return this.active
      ? this._getActiveColor()
      : 'dark-gray';
  }

  onClick(event: MouseEvent) {
    this.click.emit(event);
  }
}
