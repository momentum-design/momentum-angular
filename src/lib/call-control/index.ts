export * from './call-control.module';
export * from './call-control.component';
