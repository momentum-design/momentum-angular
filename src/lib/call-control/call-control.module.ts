import { NgModule, ModuleWithProviders } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ButtonModule } from '../button';
import { IconModule } from '../icon';
import { CallControlComponent } from './call-control.component';

@NgModule({
  imports: [CommonModule, ButtonModule, IconModule],
  declarations: [CallControlComponent],
  exports: [CallControlComponent],
})
export class CallControlModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: CallControlModule,
      providers: [],
    };
  }
}
