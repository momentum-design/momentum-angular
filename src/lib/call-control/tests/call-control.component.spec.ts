import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { Component } from '@angular/core';
import { CallControlModule } from '../call-control.module';

describe('CallControlComponent', () => {
  @Component({
    selector: 'test-app',
    template: `
      <md-call-control
        id="test-default"
        [type]="type"
        ariaLabel="For the Win"
        size=20
        [iconColor]="iconColor"
        [iconSize]="iconSize"
        (click)="onClick()"
      ></md-call-control>

      <md-call-control
        id="test-active"
        [type]="type"
        ariaLabel="For the Win"
        [active]="true"
      ></md-call-control>

      <md-call-control
        id="test-disabled"
        [type]="type"
        ariaLabel="For the Win"
        [disabled]="true"
      ></md-call-control>
    `,
  })
  class TestAppComponent {
    count = 0;
    iconColor = 'white-100';
    iconSize = 20;
    type = 'microphone-muted';
    onClick() {
      this.count++;
    }
  }

  let component: TestAppComponent;
  let fixture: ComponentFixture<TestAppComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TestAppComponent],
      imports: [CallControlModule],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TestAppComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should match snapshot', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should render one Call Control button', () => {
    const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-default');
    expect(element.nodeName).toEqual('BUTTON');
  });

  it('should handle disabled state', () => {
    const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
    expect(element.getAttribute('disabled')).toEqual('true');
  });

  it('should handle click event', () => {
    const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-default');
    element.click();
    expect(component.count).toEqual(1);
  });

  describe('tests for type prop', () => {
    it('should render activities type', () => {
      component.type = 'activities';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('activities_24');
    });

    it('should render camera type', () => {
      component.type = 'camera';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('camera_24');
    });

    it('should render camera-muted type', () => {
      component.type = 'camera-muted';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('camera-muted_24');
    });

    it('should render camera-muted type', () => {
      component.type = 'camera-muted';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('camera-muted_24');
    });

    it('should render cancel type', () => {
      component.type = 'cancel';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('cancel_24');
    });

    it('should render handset type', () => {
      component.type = 'handset';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('handset_24');
    });

    it('should render microphone-muted type', () => {
      component.type = 'microphone-muted';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('microphone-muted_24');
    });

    it('should render more type', () => {
      component.type = 'more';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('more_24');
    });

    it('should render participant-list type', () => {
      component.type = 'participant-list';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('participant-list_24');
    });

    it('should render share-screen type', () => {
      component.type = 'share-screen';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('share-screen_24');
    });

    it('should render speaker type', () => {
      component.type = 'speaker';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('speaker_24');
    });

    it('should render speaker type', () => {
      component.type = 'speaker';
      fixture.detectChanges();
      const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-disabled');
      const icon = element.querySelector('md-icon');
      expect(icon.getAttribute('ng-reflect-name')).toEqual('speaker_24');
    });
  });

  it('should handle iconSize of 10', () => {
    component.iconSize = 10;
    fixture.detectChanges();
    const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-default');
    const icon: HTMLBaseElement = element.querySelector('md-icon');
    expect(icon.getAttribute('ng-reflect-name')).toEqual('microphone-muted_10');
  });

  it('should handle iconSize of 16', () => {
    component.iconSize = 16;
    fixture.detectChanges();
    const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-default');
    const icon: HTMLBaseElement = element.querySelector('md-icon');
    expect(icon.getAttribute('ng-reflect-name')).toEqual('microphone-muted_16');
  });

  it('should handle iconColor prop', () => {
    component.iconColor = 'green-50';
    fixture.detectChanges();
    const element: HTMLBaseElement = fixture.nativeElement.querySelector('#test-default');
    const icon: HTMLBaseElement = element.querySelector('md-icon');
    expect(icon.getAttribute('ng-reflect-color')).toEqual('green-50');
  });

});
