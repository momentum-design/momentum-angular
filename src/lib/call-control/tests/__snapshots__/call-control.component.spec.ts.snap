// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CallControlComponent should match snapshot 1`] = `
<test-app
  count="0"
  iconColor={[Function String]}
  iconSize={[Function Number]}
  type={[Function String]}
>
  <button
    active="false"
    aria-label="For the Win"
    class="md-button md-call-control md-button--dark-gray md-button--20 md-button--circle"
    href=""
    id="test-default"
    md-button=""
    ng-reflect-circle="true"
    ng-reflect-color="dark-gray"
    ng-reflect-disabled="false"
    ng-reflect-ng-class="md-call-control,"
    ng-reflect-size="20"
    tabindex="0"
    type="button"
  >
    
    <span
      class="md-button__children"
      ng-reflect-ng-style="[object Object]"
      style="opacity: 1;"
    >
      <md-icon
        class="md-icon icon icon-microphone-muted_20"
        ng-reflect-color="white-100"
        ng-reflect-name="microphone-muted_20"
        style="display: inline-flex; color: rgb(255, 255, 255);"
      >
        
      </md-icon>
    </span>
  </button><button
    active="false"
    aria-label="For the Win"
    class="md-button md-call-control md-button--red md-button--56 md-button--circle"
    href=""
    id="test-active"
    md-button=""
    ng-reflect-circle="true"
    ng-reflect-color="red"
    ng-reflect-disabled="false"
    ng-reflect-ng-class="md-call-control,"
    ng-reflect-size="56"
    tabindex="0"
    type="button"
  >
    
    <span
      class="md-button__children"
      ng-reflect-ng-style="[object Object]"
      style="opacity: 1;"
    >
      <md-icon
        class="md-icon icon icon-microphone-muted_24"
        ng-reflect-color="white-100"
        ng-reflect-name="microphone-muted_24"
        style="display: inline-flex; color: rgb(255, 255, 255);"
      >
        
      </md-icon>
    </span>
  </button><button
    active="false"
    aria-label="For the Win"
    class="md-button md-call-control md-button--dark-gray md-button--56 md-button--circle md-button--disabled"
    disabled="true"
    href=""
    id="test-disabled"
    md-button=""
    ng-reflect-circle="true"
    ng-reflect-color="dark-gray"
    ng-reflect-disabled="true"
    ng-reflect-ng-class="md-call-control,"
    ng-reflect-size="56"
    tabindex="0"
    type="button"
  >
    
    <span
      class="md-button__children"
      ng-reflect-ng-style="[object Object]"
      style="opacity: 1;"
    >
      <md-icon
        class="md-icon icon icon-microphone-muted_24"
        ng-reflect-color="white-100"
        ng-reflect-name="microphone-muted_24"
        style="display: inline-flex; color: rgb(255, 255, 255);"
      >
        
      </md-icon>
    </span>
  </button>
</test-app>
`;
